Add parameterset catalogue

Peter9192 opened the pull request:
Here's an initial idea on how we could support a parameterset catalogue.

Adds
* `ewatercycle.parameter_sets.available_parameter_sets()` + `ewatercycle.parameter_sets.get_parameter_set()` as shown in api notebook
* `ewatercycle.parameter_sets.download_example_parameter_sets()` to download example parameter sets.
* System setup notebook see #120 

Changes:
*  Moved MaskMap from parameter set to `Lisflood.setup()` see #121
* In notebooks replaced SubversionCopier for parameter sets with get_parameter_set
* Applied PyCharm formatting or black to touched files
* Renamed config file in user dir from `~/.config/.ewatercycle/ewatercycle.yaml` to `~/.config/ewatercycle/ewatercycle.yaml`
* AbstractModel now uses generics so model subclasses can have their own forcing class

Removed:
* Models no longer have their own parameter set class, they all use same one

Not part of this PR:
* Remove pararmetersetdb, see #131

Peter9192 committed:
initial idea for parameterset catalogue

Peter9192 committed:
Refactor catalogue into parameter_sets module

sverhoeven committed:
Added some tests

sverhoeven committed:
Use the fixture

sverhoeven committed:
More tests

sverhoeven committed:
Even more tests

sverhoeven committed:
Make target_model optional

sverhoeven committed:
More todos

sverhoeven committed:
less todos

sverhoeven committed:
Implemented download_example_parameter_sets() + other bits and pieces

Peter9192 commented on ewatercycle/config/_validators.py:
```diff
@@ -115,36 +124,30 @@ def validate_path(value, allow_none=False):
-    'container_engine': validate_string_or_none,
-    'singularity_dir': validate_path_or_none,
-    'output_dir': validate_path_or_none,
-    'ewatercycle_config': validate_path_or_none,
```
This is a breaking change. There's a reload system for the CFG object that uses this. Do we want to get rid of that?

sverhoeven replied:
Oeps, we should write ewatercycle.yaml without `ewatercycle_config` key.

Peter9192 committed:
Merge remote-tracking branch 'origin/master' into catalogue

Peter9192 committed:
formatting

Peter9192 committed:
dont write ewatercycle_config to yaml output

Peter9192 committed:
write to previous config; add todo's

Peter9192 committed:
Refactor download example parameter sets

sverhoeven committed:
Config no longer has `.` in keys, removed dead code/test

sverhoeven committed:
Added CFG.save_to_file()

sverhoeven committed:
Move example parameter set to own files

sverhoeven committed:
Format with black

sverhoeven committed:
Added docstrings + made ParameterSet public because it is used in api notebook

sverhoeven committed:
Added tests

Peter9192 commented:
Nice progress! @sverhoeven

sverhoeven committed:
Formatting

sverhoeven committed:
Replace model specific parameter set with generic one

sverhoeven committed:
Merge remote-tracking branch 'origin/master' into catalogue

sverhoeven committed:
Added notebook for download_example_parameter_sets()

sverhoeven committed:
Moved MaskMap from parameter set to setup

Fixes #121

sverhoeven committed:
Added ParameterSet.supported_model_versions + AbstractModel._check_parameter_set()

Needed a lot of refactoring to please mypy and to have all info together for _check_parameter_set()

sverhoeven committed:
Replace SubVersionCopier with get_parameter_set()

sverhoeven committed:
Replace `Path.is_relative_to()` with Py37 compatible implementation

sverhoeven committed:
Correct sphinx links

sverhoeven commented:
@Peter9192 I think the PR is ready for review. What do you think?

sverhoeven committed:
Remove unused imports

They should be part fix for #124

sverhoeven committed:
Use Lisflood.setup{MaskMap) in notebook

Peter9192 commented on tests/models/test_lisflood.py:
```diff
@@ -2,69 +2,54 @@
-            config=root / 'settings_lat_lon-Run.xml',
-        )
+    def parameterset(self, mocked_config):
+        example_parameter_set = list(example_parameter_sets())[0]
```
Don't really like the need for this `list(...)` wrapper. Is there a neat way to solve that in the code?

sverhoeven replied:
I made `example_parameter_sets()` return a dict so you can use `example_parameter_sets()['lisflood_fraser']`to get it.

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -238,26 +244,27 @@ def _coords_to_indices(self, name: str, lat: Iterable[float], lon: Iterable[floa
             ('IrrigationEfficiency', self._get_textvar_value('IrrigationEfficiency')),
             ('PathRoot', str(self.parameter_set.directory)),
-            ('MaskMap', str(self.parameter_set.MaskMap.parent)),
+            ('MaskMap', self._get_textvar_value('MaskMap')),
```
So this `_get_textvar_value` method raises an error if `MaskMap` is not set? Is that what we want here?

sverhoeven replied:
At https://ec-jrc.github.io/lisflood-code/4_annex_input-files/ it says default is `area` for MaskMap. Added defaults for MaskMap and IrrigationEfficiency

Peter9192 commented on tests/models/test_lisflood.py:
```diff
@@ -134,6 +118,62 @@ def test_get_value_at_coords_faraway(self, model: Lisflood):
+            with patch.object(BmiClientSingularity, '__init__', return_value=None) as mocked_constructor, patch(
+                'time.strftime', return_value='42'):
+                config_file, config_dir = model.setup(
+                    MaskMap=tmp_path / 'custommask/mask.map'
```
What happens if this `MaskMap` argument is omitted (and it is not part of the `parameter_set`?

sverhoeven replied:
The model container will throw an exception during `initialize()`. I feel this is fine as the ewatercycle package should not try to solve bugs in the parameter set.

Peter9192 commented on ewatercycle/models/abstract.py:
```diff
@@ -150,3 +168,20 @@ def time_step(self) -> float:
+                             f'expected {model_name} got {self.parameter_set.target_model}')
+        if self.parameter_set.supported_model_versions == set():
+            logger.warning(f'Model expects parameter set to support version {self.version}, '
+                           f'but parameter set supports any version')
```
If the parameter set supports ("should work with"?) any version, that should be just fine, right? Why issue a warning?

Peter9192 replied:
Can we use "compatible with" instead of "x supports y"?

sverhoeven replied:
I wanted to give a warning to the user that parameter set is potentially untested with the selected model version. Maybe better to lower it to `logger.info`

Peter9192 replied:
```suggestion
        if self.parameter_set.supported_model_versions == set():
            logger.info(f'Model version {self.version} is not explicitly listed in the supported model versions of this parameter set. This can lead to compatibility issues.')
```

Peter9192 commented on ewatercycle/models/marrmot.py:
```diff
@@ -262,12 +261,13 @@ def parameters(self) -> Iterable[Tuple[str, Any]]:
+              'gamma_distribution_phi_parameter')
+
 
 class MarrmotM14(AbstractModel):
```
Should this one also be associated with MarrmotForcing?

sverhoeven replied:
Oeps, you are right, fixed it.

Peter9192 commented on tests/models/test_abstract.py:
```diff
@@ -161,25 +178,80 @@ def test_time_step(bmi, model: MockedModel):
+            target_model='mockedmodel',  # == lowered class name
+            supported_model_versions=set()
+        )
+        with caplog.at_level(logging.WARNING):
```
caplog is a pytest fixture?

sverhoeven replied:
Yep see https://docs.pytest.org/en/latest/reference/reference.html?highlight=caplog#std-fixture-caplog

Peter9192 commented on ewatercycle/models/abstract.py:
```diff
@@ -1,19 +1,36 @@
+ForcingT = TypeVar('ForcingT', bound=DefaultForcing)
+
+
+class AbstractModel(Generic[ForcingT], metaclass=ABCMeta):
```
Learned something new! Nicely done :-)

sverhoeven replied:
Thanks

Peter9192 commented on tests/parameter_sets/test_example.py:
```diff
@@ -0,0 +1,69 @@
+    CFG.reload()
+
+
+class TestGoldenPath:
```
What is this?

sverhoeven replied:
I think I started with the idea of having tests for the happy code path grouped together and other tests for when things go wrong. But seems I did not followed through. Flattened tests again

Peter9192 reviewed (commented):
Did a first pass, haven't tested or looked at the notebook yet. I have a few comments so far.

Peter9192 commented on ewatercycle/parameter_sets/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    """Lists example parameter sets that can be downloaded with :py:func:`~download_example_parameter_sets`.
+    """
+    # TODO how to add a new model docs should be updated with this part
+    return chain(
```
I'd rather have this return a dict or so. Then you could do:

```python
examples = ewatercycle.parameter_sets.example_parameter_sets()
print(examples)
> {'lisflood_fraser': ParameterSet('lisflood_fraser', ...), 'wflow_rhine_sbm_nc': ParameterSet('...'), ...}
lisflood_fraser = examples['lisflood_fraser']
lisflood_fraser.download()
... (run examples case for lisflood model)
```

sverhoeven replied:
Yep, done dict implemented and used in test, but for notebooks I would like to download all examples. This is much less verbose.

Peter9192 commented on ewatercycle/parameter_sets/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    return parametersets
+
+
+def available_parameter_sets(target_model: str = None) -> Iterable[str]:
```
`ewatercycle.parameter_sets.available_parameter_sets(target_model="lisflood")` currently returns a generator, which is not ideal for the public API (and not consistent with the example notebook). It would be nice if we can incorporate this example in one of our notebooks, and ideally it should return a list of parameterset names or a dict with them names as keys.

sverhoeven replied:
Changed it to tuple of strings as that more closely matches to api notebook.

Peter9192 commented on ewatercycle/parameter_sets/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    Returns: Names of available parameter sets on current machine.
+
+    """
+    all_parameter_sets = _parse_parametersets()
```
Can we print a nice message when there is no available parameterset? Currently it returns an empty generator...

sverhoeven replied:
Added, still needs docs explaining how to add a pararmeter set

Peter9192 commented on ewatercycle/models/abstract.py:
```diff
@@ -150,3 +168,20 @@ def time_step(self) -> float:
+                           f'but parameter set supports any version')
+        elif self.version not in self.parameter_set.supported_model_versions:
+            raise ValueError(
+                f'Parameter set is not supported with version {self.version} of model, '
```
`model = ewatercycle.models.Wflow('2020.1.13', parameters, forcing)` now returns an error about the parameter set not being compatible: `Parameter set is not supported with version 2020.1.13 of model, parameter set only supports {'2020.1.1'}`

I see two problems with this:

1. The first point of failure should be the model itself/the container. If the version is not in model.available_versions that should raise first. Only then should it proceed to check compatibility with parameterset.
2. I find the message confusing, since I'm not sure who is supposed to support what? Does the model support the parameterset? Perhaps it is clearer like this? 

```suggestion
                f'Parameter set is not compatible with version {self.version} of model, '
```

sverhoeven replied:
Added check for version against available_version, which happens before pararmeter set checks.

Peter9192 commented on ewatercycle/parameter_sets/default.py:
```diff
@@ -0,0 +1,65 @@
+
+    def __init__(
+        self,
+        name: str,
```
In the example API notebook, `id` is used instead of `name`. Maybe change it, or support both?

```py
@property
def id(self):
    return self.name
```

sverhoeven replied:
`id` is a built-in. In PyCharm I get `Shadows built-in name 'id' ` warning. I think we should stick with `name`.

Peter9192 reviewed (commented):
Tested with the notebooks. A few more comments about the functionality.

sverhoeven committed:
example_parameter_sets() return a dict instead of iterator

sverhoeven committed:
Fallback to default value when key is missing in config

sverhoeven committed:
Make MarrmotM14 a non-generic class

sverhoeven committed:
Flatten test hierarchy

sverhoeven committed:
Correct nesting

sverhoeven committed:
Raise exceptions in available_parameter_sets and return tuple

sverhoeven committed:
Update ewatercycle/models/abstract.py

Co-authored-by: Peter Kalverla <peter.kalverla@gmx.com>

sverhoeven committed:
Check model.version against model.available_versions

sverhoeven committed:
List class var before object vars

sverhoeven committed:
Merge branch 'catalogue' of github.com:eWaterCycle/ewatercycle into catalogue

sverhoeven committed:
Remove duplicate bmi in docs

sverhoeven commented:
@Peter9192 could you have another look?

sverhoeven committed:
Fix tests

Peter9192 commented on ewatercycle/models/lisflood.py:
```diff
@@ -259,11 +261,11 @@ def parameters(self) -> Iterable[Tuple[str, Any]]:
+        # TODO fix issue #84
         parameters = [
-            ('IrrigationEfficiency', self._get_textvar_value('IrrigationEfficiency')),
+            ('IrrigationEfficiency', self._get_textvar_value('IrrigationEfficiency'), 0.75),
```
So here we pretend there's a default, even if it's not set in the model config? And then the model itself will fail?

sverhoeven replied:
Yes, it will fail when area.map or area.nc file does not exist

sverhoeven replied:
Reverted having a default in the model, will raise KeyError when MaskMap or IrrigationEfficiency can not be found in config.
An bmi model will return error when file from config does not exist

Peter9192 reviewed (commented):
Almost there. I made a suggestion for that warning message. Also two minor points about the notebooks:

- `download_example_parameter_sets` fails when one of the parameterset dirs already exists. Might be nice to have a 'skip_existing' and 'force_overwrite' option. Now, the CFG is also not updated, and my CFG.reload results in an old config file. 
- can we change `id` to `doi` in the example API notebook as well, and maybe execute all the commands in the cell that's about the parameter sets, instead of having fake output in comments?

sverhoeven committed:
Update ewatercycle/models/abstract.py

Co-authored-by: Peter Kalverla <peter.kalverla@gmx.com>

sverhoeven committed:
Replace id with name

sverhoeven committed:
Add skip_existing arg to download

sverhoeven committed:
Revert "Fallback to default value when key is missing in config"

This reverts commit f776fce4e581999c527f83022b78557871d98a02.

sverhoeven committed:
Merge remote-tracking branch 'origin/master' into catalogue

Also changed MaskMap to str type

sverhoeven committed:
Fix test

sverhoeven committed:
Fix mypy warning
